src/Utils/DPop.php

```diff
@@ -76,109 +70,146 @@ public function getWebId($request) {
 	}
 
 	/**
-	 * Returns the "kid" from the "jwk" header in the DPoP token.
-	 * The DPoP token must be valid.
-	 *
-	 * @param string $dpop The DPoP token
-	 * @param ServerRequestInterface $request Server Request
-	 *
-	 * @return string          the "kid" from the "jwk" header in the DPoP token.
-	 *
-	 * @throws RequiredConstraintsViolated
+	 * kept for backwards compatability
+	 * note: the "kid" value is not guaranteed to be a hash of the jwk
+	 * so to compare a jkt, calculate the jwk thumbprint instead
+	 * @param  string $dpop    The DPoP token, raw
+	 * @param  ServerRequestInterface $request Server Request
+	 * @return string          The "kid" from the "jwk" header
 	 */
 	public function getDpopKey($dpop, $request) {
-		$kid = '';
-
 		$this->validateDpop($dpop, $request);
 
-		// 1.  the string value is a well-formed JWT,
 		$jwtConfig = Configuration::forUnsecuredSigner();
 		$dpop = $jwtConfig->parser()->parse($dpop);
 		$jwk  = $dpop->headers()->get("jwk");
 
-		if (isset($jwk['kid'])) {
-			$kid = $jwk['kid'];
+		if (isset($jwk['kid']) === false) {
+			throw new InvalidTokenException('Key ID is missing from JWK header');
+		}
+
+		return $jwk['kid'];
+	}
+
+	/**
+	 * RFC7638 defines a method for computing the hash value (or "digest") of a JSON Web Key (JWK).
+	 *
+	 * The resulting hash value can be used for identifying the key represented by the JWK
+	 * that is the subject of the thumbprint.
+	 *
+	 * For instance by using the base64url-encoded JWK Thumbprint value as a key ID (or "kid") value.
+	 *
+	 * @see https://www.rfc-editor.org/rfc/rfc7638
+	 *
+	 * The thumbprint of a JWK is created by:
+	 *
+	 * 1. Constructing a JSON string (without whitespaces) with the required keys in alphabetical order.
+	 * 2. Hashing the JSON string using SHA-256 (or another hash function)
+	 *
+	 * @param string $jwk The JWK key to thumbprint
+	 * @return string the thumbprint
+	 * @throws InvalidTokenException	 
+	 */
+	public function makeJwkThumbprint($jwk) {
+		if (!$jwk || !isset($jwk['kty'])) {
+			throw new InvalidTokenException('JWK has no "kty" key type');
 		}
+		if (!in_array($jwk['kty'], ['RSA','EC'])) {
+			throw new InvalidTokenException('JWK "kty" key type value must be one of "RSA" or "EC", got "'.$jwk['kty'].'" instead.');
+		}
+		if ($jwk['kty']=='RSA') {
+			if (!isset($jwk['e']) || !isset($jwk['n'])) {
+				throw new InvalidTokenException('JWK values do not match "RSA" key type');
+			}
+			$json = vsprintf('{"e":"%s","kty":"%s","n":"%s"}', [
+				$jwk['e'],
+				$jwk['kty'],
+				$jwk['n'],
+			]);
 
-		return $kid;
+		} else { //EC
+			if (!isset($jwk['crv']) || !isset($jwk['x']) || !isset($jwk['y'])) {
```

Comment:
No need for an or `||`, as `isset` accepts multiple parameters:

```
if ( ! isset($jwk['crv'], $jwk['x'], $jwk['y'])) {
```

Reply:
done